stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,415 @@
+   * layer. It will not block while the message is being sent on the wire and returning true does
+   * not guarantee that the server gets the message.
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
+   * control.  Furthermore, the server closing the stream will only be identified after the last
+   * sent value is read.
+   * </p>
+   *
+   * @param request Message to send to the server
+   * @return true if the request is sent to stream, false if skipped
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public boolean write(ReqT request) throws InterruptedException {
+    try {
+      return write(request, Integer.MAX_VALUE, TimeUnit.DAYS);
+    } catch (TimeoutException e) {
+      throw new RuntimeException(e); // should never happen
+    }
+  }
+
+  /**
+   * Send a value to the stream for sending to server, wait if necessary for the grpc stream to be
+   * ready up to specified timeout.
+   * <br>
+   * If write is not legal at the time of call, immediately returns false
+   * <br><br>
+   * <p>
+   * <b>NOTE:  </b>This method will return as soon as it passes the request to the grpc stream
+   * layer. It will not block while the message is being sent on the wire and returning true does
+   * not guarantee that the server gets the message.
+   * </p>
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
+   * control.  Furthermore, the server closing the stream will only be identified after the last
+   * sent value is read.
+   * </p>
+   *
+   * @param request Message to send to the server
+   * @param timeout How long to wait before giving up.  Values &lt;= 0 are no wait
+   * @param unit A TimeUnit determining how to interpret the timeout parameter
+   * @return true if the request is sent to stream, false if skipped
+   * @throws TimeoutException if write does not become ready before the specified timeout expires
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public boolean write(ReqT request, long timeout, TimeUnit unit)
+      throws InterruptedException, TimeoutException {
+    executor.drain();
+
+    if (getClosedStatus() != null && !getClosedStatus().isOk()) {
+      throw getClosedStatus().asRuntimeException();
+    }
+
+    if (writeClosed) {
+      throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+    } else if (!isWriteLegal()) {
+      return false;
+    }
+
+    boolean writeDone = false;
+
+    long start = System.nanoTime();
+    long duration = unit.toNanos(timeout);
+
+    while (!writeDone && !writeClosed) {
+      if (call.isReady()) {
+        call.sendMessage(request);
+        writeDone = true;
+      } else if (getClosedStatus() == null) {
+        waitAndDrainExecutorOrTimeout(start, duration);
```

Comment:
This while loop logic is  overly complicated and im wondering if it is necessary. Can we do this:
```
while (!writeDone && !writeClosed && getClosedStatus()!=null) {
  if (call.isReady()) {
  ...
  } else {
  waitAndDrain
  }
}
```
Then the if block before in L209  may also be removed.